src/script/lua_api/l_object.cpp

```diff
@@ -1701,43 +1701,183 @@ int ObjectRef::l_hud_get_hotbar_selected_image(lua_State *L)
 	return 1;
 }
 
-// set_sky(self, bgcolor, type, list, clouds = true)
+// set_sky(self, {bgcolor=, type=, param=, sky_color=, direc_fog, base_fog})
 int ObjectRef::l_set_sky(lua_State *L)
 {
 	NO_MAP_LOCK_REQUIRED;
 	ObjectRef *ref = checkobject(L, 1);
 	RemotePlayer *player = getplayer(ref);
-	if (player == NULL)
+	if (!player)
 		return 0;
 
-	video::SColor bgcolor(255,255,255,255);
-	read_color(L, 2, &bgcolor);
+	// Check for ColorSpec tables:
+	bool is_colorspec = false;
+	if (lua_istable(L, 2)) {
+		// Iterate over the whole table in case of missing ColorSpec keys:
+		// Only update the value if we know we have a valid ColorSpec table.
+		lua_getfield(L, 2, "r");
+		if (!is_colorspec) 
+			is_colorspec = lua_isnumber(L, -1);
+		lua_getfield(L, 2, "g");
+		if (!is_colorspec)
+			is_colorspec = lua_isnumber(L, -1);
+		lua_getfield(L, 2, "b");
+		if (!is_colorspec)
+			is_colorspec = lua_isnumber(L, -1);
```

Comment:
Please use `read_ARGB8` here or move this `ColorSpec`-specific code to `c_converter.cpp`.

Reply:
Probably worth moving to c_converter.cpp as this is a read only, non stack-popping operation.